include/swift/AST/TypeCheckRequests.h

```diff
@@ -1095,7 +1095,7 @@ class InferredGenericSignatureRequest :
     public SimpleRequest<InferredGenericSignatureRequest,
                          GenericSignature (ModuleDecl *,
                                             GenericSignatureImpl *,
-                                            GenericParamList *,
+                                            PointerUnion<Decl *, GenericParamList *>,
```

Comment:
I'm a bit confused why there's a Decl here. Can you store the where clause directly here? Also maybe define a type alias for this since it appears in multiple places.

Reply:
I had to because WhereClauseOwner accepts a `Decl` rather that a `TrailingWhereClause`
https://github.com/apple/swift/blob/217241f5050327e118f2119d3f2df671b054ec55/include/swift/AST/TypeCheckRequests.h#L369-L372

I think this is because trailing requirements get subsumed by `GenericParamList` during parsing if the decl is generic, so sometimes (`protocol`, `associatedtype`) we need to look for requirements in the where clause and sometimes -- in the gen param list.

Reply:
We could pass around a `GenericContext` instead if `AssociatedTypeDecl` -- the edge case -- were one :-/

Reply:
We can also use `llvm::PointerUnion<GenericParamList *, GenericContext *, AssociatedTypeDecl *, SpecializeAttr *>` instead of `llvm::PointerUnion<GenericParamList *, Decl *, SpecializeAttr *>` for `WhereClauseOwner` as a temporary solution and a way to highlight the edge case problem.

Reply:
Defined a type alias and passing a `GenericContext` now.